### PR TITLE
Make TokenForwarder and Switchboard  target capabilities retrievable

### PR DESCRIPTION
**Pull request by austinkline**
Closes: #???

## Description

Add a few helper methods that will make it easier to validate underlying receiver capabilities in the `TokenForwarder` and `FungibleTokenSwitchboard` contracts. Without this, it is much harder to know what we are looking at under the hood, which means that malicious users could setup any number of roundabout ways to make receivers actually invalid (like an infinite loop of receivers that will drain all gas in a transaction)

Made a brief diagram to showcase this problem, it applies to both TokenForwarding and the Switchboard:
flow-ft

In any case, without some mechanism to investigate what is being referenced in the forwarder or switchboard resources, there isn't a way on-chain to ensure that what we are sending through can actually reach its destination. Because of that, I can "poison" the chain of resources anywhere down the line to prevent a token vault from being able to be sent. 

With these editions, we can make some form of validator that only permits a maximum depth, and rejects the deposit if it goes beyond that. This will allow marketplaces to be defensive and reject vault distribution if it goes too far, or if it has the wrong type along its chain (a non forwarder or switchboard)
______

For contributor use:

- [ ] Targeted PR against `master` branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/flow-ft/blob/master/CONTRIBUTING.md#styleguides).
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels 

- [ ] Update the version in package.json when there is a change in the smart contracts

### CODE REVIEW

**Comment by austinkline** [comment]
Here is an example contract that could use these changes:
https://gist.github.com/austinkline/f7f80ec5ea04ef168de59a0778ddc438

**NOTE: This is not tested, it is meant to demonstrate the problem and why a solution like this is necessary**

**Review by joshuahannan — APPROVED** [review]
Really great suggestion, and well implemented. Are you able to add a test for these? I can try to get others to look at it

**Comment by austinkline** [comment]
> Really great suggestion, and well implemented. Are you able to add a test for these? I can try to get others to look at it

Should have some time next week to add them 🙏

File: contracts/utility/TokenForwarding.cdc
Comment (austinkline) [inline contracts/utility/TokenForwarding.cdc:57]: Thinking through this.. Is there a scenario where this borrow panics? Isn't that what happens after secure cadence if the completely wrong type is underneath?

In that case, we will need to check first, and then return the borrow result afterwards

Comment (joshuahannan) [reply contracts/utility/TokenForwarding.cdc:57]: that is a good point. I think you might be right. better to check beforehand

**Comment by bluesign** [comment]
there is also good old cyclic link: (panics even on check)
```
acct.link<&AnyStruct>(/public/deniz, target: /public/bluesign)
acct.link<&AnyStruct>(/public/bluesign, target: /public/deniz)
```

**Comment by bjartek** [comment]
is it possible to have this feature merged? We want to use Switchboard as royalty receiver in Doodles to handle royalties on both dapper and non dapper wallet, but I do not want to break it for @austinkline, and he needs this feature.

**Comment by joshuahannan** [comment]
There are still a few unresolved comments so we can't merge it yet unfortunately

**Comment by bluesign** [comment]
I am worried a bit on this `safeBorrow` , maybe we can add some checks in adding capability? otherwise if someone accidentally adds `&Vault` as receiver, borrower can take a reference to all vault.

**Comment by joshuahannan** [comment]
The safe borrow method returns a restricted reference `&{FungibleToken.Receiver}`, so it can't return a vault reference, just a receiver reference

**Comment by bluesign** [comment]
> The safe borrow method returns a restricted reference `&{FungibleToken.Receiver}`, so it can't return a vault reference, just a receiver reference

ah true, I though somehow it is just making `borrow()`

**Comment by bluesign** [comment]
```
    pub fun safeBorrowByType(type: Type): &{FungibleToken.Receiver}? {
             if !self.checkReceiverByType(type: type) {
                 return nil
             }

             return self.receiverCapabilities[type]!.borrow()
         }
```

this one is problematic

**Comment by joshuahannan** [comment]
@bluesign How is that problematic? It still only returns receiver references

**Comment by bjartek** [comment]
any ETA on when this can be merged? It is needed.

**Comment by joshuahannan** [comment]
We still need to figure out if @bluesign's comment is an issue before we can merge it. I don't believe it is a problem, but I just want to double check. In addition, there are still no tests in this PR, so I'm not comfortable merging it until there are some tests written.

**Comment by bjartek** [comment]
```import FungibleToken from "../contracts/standard/FungibleToken.cdc"
import FungibleTokenSwitchboard from "../contracts/standard/FungibleTokenSwitchboard.cdc"
import FUSD from "../contracts/standard/FUSD.cdc"

transaction() {

	prepare(acct: AuthAccount) {
		// Get a Receiver reference for the Dapper account that will be the recipient of the forwarded DUC and FUT

		//FUSD
		let fusdReceiver = acct.getCapability<&{FungibleToken.Receiver}>(/public/fusdReceiver)
		if !fusdReceiver.check() {
			let fusd <- FUSD.createEmptyVault()
			acct.save(<- fusd, to: /storage/fusdVault)
			acct.link<&FUSD.Vault>( /public/fusdReceiver, target: /storage/fusdVault)
		}
		let switchboard <- FungibleTokenSwitchboard.createSwitchboard()
		switchboard.addNewVault(capability: fusdReceiver)


		if let typ = switchboard.safeBorrowByType(type: Type<@FUSD.Vault>()) {

			let vault = typ as! &FUSD.Vault
			//compile error

		}
		destroy switchboard
	}
}
```

**Comment by bluesign** [comment]
ah yeah probably I remember wrong somehow, it works with `let fusdReceiver = acct.getCapability<&FUSD.Vault>(/public/fusdReceiver)` though.

**Comment by austinkline** [comment]
> any ETA on when this can be merged? It is needed.

Yeah, the ball is in my court here I just haven't had the time to leave flowty work to finish this out with tests and some small alterations (doing a check before a borrow)

**Comment by austinkline** [comment]
@bluesign hadn't thought about the downcasting issue, we could probably throw together a quick script to make sure what you're describing isn't possible, though. My understanding is that since borrow is has T set to `&{FungibleToken.Receiver}`, we should be good but definitely best to confirm

**Comment by bluesign** [comment]
@austinkline yeah my example was little bad; 

I think situation is more like this: 

- I have a vault linked as `&FlowToken.Vault to `/private/denizVault`
- I can get capability to this: `getCapability<&FlowToken.Vault>(/private/denizVault)`
- `switchboard.addNewVault` will accept this as `&FlowToken.Vault` is also `&AnyResource{FT.Receiver}`
- then `switchboard.safeBorrowByType` will borrow this as `&FlowToken.Vault`
- it will return `&FlowToken.Vault` ( return type will be ok as it is also `&AnyResource{FT.Receiver}` ) 
- then who got this, can cast it to `&FlowToken.Vault`

I mean it is not a big deal, but also it is unexpected.

**Comment by alilloig** [comment]
Wow @bluesign you have a twisted mind (in the best sense ofc ;) ), I love it!!  Honestly I had never thought of that, @turbolent could you please help us confirm that the specified type on a returning value is checking a conformance rather than restricting the type?

@austinkline would you mind if I made an issue and a new PR out of this adding some tests? I have done the rest of the tests for the switchboard so I should be fine. If you don't have any objection I'll do it on Monday so @bjartek can have it asap

**Comment by austinkline** [comment]
> @austinkline would you mind if I made an issue and a new PR out of this adding some tests? I have done the rest of the tests for the switchboard so I should be fine. If you don't have any objection I'll do it on Monday so @bjartek can have it asap

Feel free!

**Comment by joshuahannan** [comment]
I'm closing this because it is included in https://github.com/onflow/flow-ft/pull/125 now. That could use some more reviews if anyone has time!

**Pull request closed without merging**
